File: AddExpressionOperator.java
/*
Time Complexity : Exponential
Space Complexity : O(digits in integer)
Idea : We have 4 choices either we add (+,-,*) or no operator. We can never add operator when position is 0 so
we always skip there for other positions we take one of the three operators since the precedence of multiplication
has to be accounted we keep track of last operation that was performed and revert our value by using the formula.
We backtrack for strinbuilder by keeping track of the length before starting the recursion and the again settingthe old length.
*/
class Solution {
    List<String> ans;
    public List<String> addOperators(String num, int target) {
        ans = new ArrayList<>();
        if(num == null || num.length() == 0){
            return ans;
        }
        helper(num,new StringBuilder(""),target,0,0,0);
        return ans;
    }
    
    public void helper(String num,StringBuilder path,int target,int pos,long curr,long lastOp){
        if(pos == num.length()){
            if(curr == target){
             ans.add(path.toString());
             return;
            }
        }
        for(int i=pos;i<num.length();i++){
           if(i > pos && num.charAt(pos) == '0') break;
            String valS = num.substring(pos,i+1);
            long val = Long.parseLong(valS);
            int len = path.length();
            if(pos == 0){
                helper(num,path.append(valS),target,i+1,val,val);
                path.setLength(len);
            }else{
                helper(num,path.append("+").append(valS),target,i+1,curr+val,val);
                path.setLength(len);
                helper(num,path.append("-").append(valS),target,i+1,curr-val,-val);
                path.setLength(len);
                helper(num,path.append("*").append(valS),target,i+1,curr-lastOp + lastOp*val,lastOp*val);
                path.setLength(len);
            }
        }
    }
   
}

File: AddExpressionOperator.py
# // Time Complexity : O(3^n)
# // Space Complexity : O(3^n)
# // Did this code successfully run on Leetcode : Yes
# // Three line explanation of solution in plain english 

# // Your code here along with comments explaining your approach
class solution:
    def addOperators(self, num, target):
        self.result = []

        '''
        the parameters of our helper function are :

        num -> the original string
        currSofar --> the total evaluation of the expresiion uptil now
        path --> string of current path taken incldung numbers and operators
        prev --> this is basically the tail of the previous operation that we conducted on currSofar

        idx --> this is the index that points the element on in our string.
        '''
        self.helper(self,num,  0, "", target, 0, 0 )
        return self.result
    def helper(self, num, currSofar, path, target, prev, idx):
        '''
        This is a tricky backtracking question as it involes multiple parts. 
        '''
        ##base

        '''
        This is our base case, we see if our index has reached the end of string. and check if currSofar is equal to target. if yes then add path to
        the result
        '''
        if idx == len(num)-1:
            if currSofar == target:
                self.result.append(path[:])
                return 
            else:
                return 

        ##logic

        '''
        here in the for loop we go through from the given idx to end of string.

        we need to remember that idx is stationary and i only moves. idx is nothing but pivot
        '''
        
        for i in range(idx, len(num)):
            '''
            this takes care of leading zeroes. Since leading zeroes do not contribute to a number. we avoid them

            '''
            if i != idx and num[IndexError]=="0":
                continue
            
            '''
            we calculate the current number from the string . we take a substring from to idx to i

            '''
            curr = num[idx:i+1]
            curr = int(curr)
            '''
            we check if the idx is at start of string. if yes then we will not have any tail or prev or we will not be adding any operators to this as this will
            just contain numbers. 
            '''
            if idx == 0:
                self.helper(self, num, curr, path+str(curr), target, curr, i+1 )

            else:
                '''
            if the idx is greater than 0 then we can do our operations of "+"
            , "- ,"*  . we add/sub the curr to currSofar and put them as prev and move to next idx


            in case of multiplication. we follow bodmas rule, so we subtract the tail from currsofar and add the product of curr and prev to currSofar and pass the product as tail/prev 

            and move to next pointer.  
                '''
                self.helper(num, currSofar+curr, path+"+"+str(curr) , target, curr, i+1 )

                self.helper(num, currSofar-curr, path+"-"+str(curr), target, -curr, i +1)

                self.helper(num, currSofar-prev +prev*curr, path+"*"+str(curr), target, prev*curr, idx+1)


